### ISSUE TITLE
SFTTrainer finishes before max_steps

### ISSUE THREAD

**Issue by liechtym** [lvwerra/trl#450]
I'm using SFTTrainer to finetune open_llama_7b with Qlora. 

My problem is the trainer finishes early, often before the halfway point. With 10000 max steps, it finishes at around 3500. My training dataset has 12,667 rows. There is no error, it just finishes.

Also the Epoch stats seem really off.

` [ 3230/10000 2:20:47 < 4:55:16, 0.38 it/s, Epoch 11.26/13]`

If I'm using a batch size of 4, then wouldn't 1 epoch be 12,667/4 = 3167 steps. Why does it say 11.26 epochs? And further, why does it say that it's at 11.26/13 epochs when it should only be 1/3 of the way done with training?

If this is an issue with transformers, not trl, let me know.

Dataset summary:

```
DatasetDict({
    train: Dataset({
        features: ['instruction', 'context', 'response', 'category'],
        num_rows: 12667
    })
    test: Dataset({
        features: ['instruction', 'context', 'response', 'category'],
        num_rows: 1408
    })
})
```

Installs:
```
!pip install -q -U bitsandbytes
!pip install -U -q git+https://github.com/huggingface/transformers@main
!pip install -q -U git+https://github.com/huggingface/peft.git
!pip install -q -U git+https://github.com/huggingface/accelerate.git
!pip install -q -U git+https://github.com/huggingface/datasets.git
!pip install -q -U git+https://github.com/lvwerra/trl.git
!pip install sentencepiece
```

Code:
```
import torch
import transformers
from peft import LoraConfig, get_peft_model
from transformers import AutoModelForCausalLM, BitsAndBytesConfig, AutoTokenizer, LlamaTokenizerFast
from peft import PeftModel, PeftConfig, get_peft_model

bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_use_double_quant=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_compute_dtype=torch.bfloat16
)

base_model = AutoModelForCausalLM.from_pretrained(
    model_id,
    quantization_config=bnb_config,
    device_map={"": 0}
)

peft_model = PeftModel.from_pretrained(model=base_model, model_id=save_model_path, quantization_config=bnb_config, is_trainable=True)

lora_config = LoraConfig.from_pretrained(save_model_path)

tokenizer = LlamaTokenizerFast.from_pretrained(save_model_path)

import transformers

from trl import SFTTrainer

trainer = SFTTrainer(
    peft_model,
    train_dataset=formatted_dataset["train"],
    eval_dataset=formatted_dataset["test"],
    args=transformers.TrainingArguments(
        per_device_train_batch_size=4,
        gradient_accumulation_steps=4,
        learning_rate=2e-4,
        max_steps=10000,
        output_dir="./output",
        optim="paged_adamw_8bit",
        logging_steps=100, 
        fp16=True
    ),
    tokenizer=tokenizer,
    peft_config=lora_config,
    dataset_text_field="text",
    max_seq_length=512,
    packing=True,
    
)
peft_model.config.use_cache = False

trainer.train()
```

**Comment by younesbelkada** [comment]
Hi @liechtym 
Thank you very much for the detailed issue.
I just made https://github.com/lvwerra/trl/pull/452 that should resolve your problem.
In fact what you are observing is expected as you are using a packed dataset. That dataset will try to create the maximum possible number of samples by packing sequences together until they reach `max_seq_len`. 
However there is a bug. 
If a user explictly passes `max_steps` to the training arguments, the trainer should in fact train the model for that many steps, regardless of the potential number of epochs it will reach. 
As a quick fix on your side, you could pass `infinite=True` to `SFTTrainer` and it should temporarly fix the issue.

**Comment by wjfu99** [comment]
Hi @younesbelkada 
Thanks for your revision, but there seems still have some problems, I still meet the same issus as the trl has been updated to the latest 0.7.1. Thus, I have the following questions:
1. In my understanding, the training process is still accomplish the`num_train_epochs` declare in the training arguments , it just a larger step length assined to tqdm, which causes the "early-stopping" of training. Do I understand it correctly?
2. Is there still need to mannually set the `max_steps` to avoid this problem? I mean a more elegant and more formal solution?
3. I believe the max_steps should be set to datasets_len / GPU_nums / gradient_accumulation_steps * epoch_num, right?

Here is a screenshot, the progess bar is still longer than expected since I set `packing=True` and `max_seq_length=1000`:
![image](https://github.com/huggingface/trl/assets/57850011/b584cfc4-3c45-40d9-9fd4-9ef9c3050f06)
